enhancements/microshift/kubernetes-for-industrial-edge.md

```diff
@@ -0,0 +1,435 @@
+   network bandwidth.
+5. As an device administrator, I can rely on the system to rollback to
+   a good state if an upgrade fails.
+6. As an device administrator, I have control over the operating
+   system running under MicroShift.
+
+### Goals
+
+* Describe an approach to providing a version of Kubernetes compatible
+  with the OpenShift APIs that it includes and that meets the
+  constraints described in the Motivation section.
+* Identify the minimal set of APIs that will be included and the
+  components needed to support them.
+* Design a system that can work autonomously, without requiring
+  external orchestration tools for deployment.
+* Treat MicroShift as an application running on RHEL, with minimal
+  privileges.
+* Design a system that can be configured by someone familiar with
+  Linux, without Kubernetes expertise.
+* Follow the [MicroShift design
+  principles](https://github.com/openshift/microshift/blob/main/docs/design.md)
+
+### Non-Goals
+
+* We are not trying to fit all of OpenShift into the target hardware
+  platforms.
+* We are not trying to make MicroShift self-managing, so no core
+  OpenShift operators will be included.
+* We do not plan to support multi-node clusters.
+* The details of the default CSI driver are left to a future
+  enhancement.
+* The details of the default CNI driver are left to a future
+  enhancement.
+* Use cases beyond industrial applications, such as developer
+  environments, are out of scope for now.
+* The design of MicroShift relies heavily on features of RHEL for Edge
+  and rpm-ostree. Those features are referenced here, but their
+  implementation is not described in detail.
+
+## Proposal
+
+To meet the resource consumption constraints (RAM and CPU), MicroShift
+is designed as an all-in-one binary for the components that are
+required to launch a kubernetes control plane, combined with
+components running in containers to provide runtime services after the
+control plane is up.
+
+Today the all-in-one binary includes:
+
+* etcd
+* kubernetes API server
+* kubernetes controller manager
+* kubernetes scheduler
+* OpenShift API server
+* OpenShift controller manager
+* OpenShift SCC manager
+
+The pod-based components include:
+
+* CoreDNS
+* service CA operator
+* CSI driver
+* CNI driver
+
+To allow deployers to customize and manage the OS, MicroShift will run
+on [RHEL for
+Edge](https://www.redhat.com/en/resources/meet-workload-demands-edge-computing-datasheet),
+which is designed for field-deployed devices and use cases.
+MicroShift will take advantage of the existing RHEL for Edge work by
+running as an application managed by systemd and using as few
+privileges as possible. Deployers will use standard RHEL for Edge
+tools such as RPM, rpm-ostree, and Image Builder to construct system
+images and distribute software updates. Deployers will be responsible
+for configuring the operating system components and anything that runs
+on the device outside of MicroShift.
```

Comment:
Possibly not worth drilling down to this level of detail in the enhancement, and pulling in [this][1]:

> * MicroShift provides a small, optional set of infrastructure services to support common use cases and reuses OpenShift's container images for these:
>   * openshift-dns, openshift-router, service-ca, local storage provider

which leaves me a bit confused about where MicroShift stands vs. Route DNS.  Touching some external DNS config to declare the requested domain name or wildcard entry would be something outside of the MicroShift device, so maybe that's the deployer's responsibility?  Do we create a DNSRecord with no controller to help drive that action, and expect the deployer to be the controller?  Do we not define a DNSRecord and... I'm not familiar enough with Route handling to understand if/how it collects feedback on how DNS is going.  Maybe we use [the external DNS operator scaffolding][2]?

[1]: https://github.com/openshift/microshift/blame/389794d8d9eae4e6b08b5472f063c57622d0c0f8/docs/design.md
[2]: https://docs.openshift.com/container-platform/4.10/networking/external_dns_operator/understanding-external-dns-operator.html

Reply:
Normally you would get no DNS entry pointing to the MicroShift instance, if they want to expose a dashboard to the internet for example, may be. But then it's up to the customer to manage that infrastucture (Regular DNS server by themselves, their own dyndns, ...) but it's out of scope for MicroShift.

Again, if they want to expose such dashboard on a LAN it,s their responsibility to create a mapping on the LAN DNS to the MicroShift instance, and probably have a static DHCP assignment to such instance.

Reply:
It probably helps clarifying that in many use cases device admins have no influence over network services like DNS, DHCP, etc.: No way to add entries to DNS, no way to allocate a CIDR or even a second IP address, etc. The device must behave like Alexa or a printer in a home network. Or it may be connected via an 4G network dongle and the service provider assigns DNS, IP, etc.